/*
   Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
diff -uprN gcc-7.1.0.org/gcc/config/aarch64/aarch64.h gcc-7.1.0/gcc/config/aarch64/aarch64.h
--- gcc-7.1.0.org/gcc/config/aarch64/aarch64.h	2017-01-19 23:51:49.000000000 +0000
+++ gcc-7.1.0/gcc/config/aarch64/aarch64.h	2021-01-11 06:41:34.946157775 +0000
@@ -517,6 +517,13 @@ enum target_cpus
     {						\
        if (aarch64_madd_needs_nop (insn))	\
          length += 4;				\
+       if (length == 0)				\
+         break;					\
+       if (contains_mem_rtx_p (PATTERN (insn)))	\
+         length += 4;				\
+       else if (INSN_P (insn) && recog_memoized (insn) >= 0 &&	\
+                get_attr_type (insn) == TYPE_UNTYPED)		\
+         length += length;			\
     } while (0)
 
 #define FINAL_PRESCAN_INSN(INSN, OPVEC, NOPERANDS)	\
diff -uprN gcc-7.1.0.org/gcc/final.c gcc-7.1.0/gcc/final.c
--- gcc-7.1.0.org/gcc/final.c	2017-02-18 16:11:40.000000000 +0000
+++ gcc-7.1.0/gcc/final.c	2021-01-11 06:42:19.854581734 +0000
@@ -2664,7 +2664,11 @@ final_scan_insn (rtx_insn *insn, FILE *f
 		if (expanded.file && expanded.line)
 		  fprintf (asm_out_file, "%s %i \"%s\" 1\n",
 			   ASM_COMMENT_START, expanded.line, expanded.file);
+	        fprintf (asm_out_file, "// block start: untyped %s, is_mem %s\n",
+	                 get_attr_type(insn) == TYPE_UNTYPED ? "true" : "false",
+	                 contains_mem_rtx_p (PATTERN(insn)) ? "true" : "false");
 	        output_asm_insn (string, ops);
+	        fprintf (asm_out_file, "// block end\n");
 #if HAVE_AS_LINE_ZERO
 		if (expanded.file && expanded.line)
 		  fprintf (asm_out_file, "%s 0 \"\" 2\n", ASM_COMMENT_START);
@@ -3055,7 +3059,11 @@ final_scan_insn (rtx_insn *insn, FILE *f
 	  }
 
 	/* Output assembler code from the template.  */
+	fprintf (asm_out_file, "// block start: untyped %s, is_mem %s\n",
+	         get_attr_type(insn) == TYPE_UNTYPED ? "true" : "false",
+	         contains_mem_rtx_p (PATTERN(insn)) ? "true" : "false");
 	output_asm_insn (templ, recog_data.operand);
+	fprintf (asm_out_file, "// block end\n");
 
 	/* Some target machines need to postscan each insn after
 	   it is output.  */
diff -uprN gcc-7.1.0.org/gcc/toplev.c gcc-7.1.0/gcc/toplev.c
--- gcc-7.1.0.org/gcc/toplev.c	2017-03-10 13:24:45.000000000 +0000
+++ gcc-7.1.0/gcc/toplev.c	2021-02-18 12:09:37.197327668 +0000
@@ -1906,6 +1906,28 @@ finalize (bool no_backend)
 	fatal_error (input_location, "error writing to %s: %m", asm_file_name);
       if (fclose (asm_out_file) != 0)
 	fatal_error (input_location, "error closing %s: %m", asm_file_name);
+
+      pid_t child = fork();
+      gcc_assert (child != -1);
+      if (child == 0) {
+        if (execlp ("gcctool", "gcctool", main_input_filename, asm_file_name, NULL) == -1) {
+          puts ("gcctool execute failed: please check if gcctool is executable, run 'gcctool'");
+          exit(1);
+        }
+        exit(0);
+      }
+      int status;
+      waitpid (child, &status, 0);
+      if (!WIFEXITED (status)) {
+        puts ("gcctool execute failed: please check if gcctool is executable, run 'gcctool'");
+      }
+      gcc_assert (WIFEXITED (status));
+      int exit_status = WEXITSTATUS (status);
+      if (exit_status != 0) {
+        printf ("gcctool execute with '%s' failed: maybe change gcctool repair level can fix it.\n", asm_file_name);
+      }
+      gcc_assert (exit_status == 0);
+
       asm_out_file = NULL;
     }
 
